src/Network/GRPC/MQTT/Proto.hs

```diff
@@ -0,0 +1,672 @@
+  | -- | 'OptionValueTypeMismatch' errors represents a mismatch between the
+    -- type expected for an option and observed value the option was set to.
+    --
+    -- * 'DatumRep' expected type of value for the option.
+    --
+    -- * 'DotProtoOption' is the observed option.
+    --
+    -- @
+    -- extend google.protobuf.MessageOptions {
+    --   // The option "my_option" is expected to be a string.
+    --   optional string my_option = 51234;
+    -- }
+    --
+    -- message MyMessage {
+    --   // Setting "my_option" to an integer value would produce a type error.
+    --   option (my_option) = 1336;
+    -- }
+    -- @
+    OptionValueTypeMismatch DatumRep DotProtoOption
+  deriving stock (Data, Eq, Generic, Ord, Show, Typeable)
+
+-- | @since 0.1.0.0
+instance Exception ProtoOptionError where
+  displayException (ProtoOptionAlreadySet idt) =
+    "the option " ++ showProtoId idt ++ " was already set."
+  displayException (OptionValueTypeMismatch rep opt) =
+    let idt = DotProto.dotProtoOptionIdentifier opt
+        val = DotProto.dotProtoOptionValue opt
+     in "expected the value set for option "
+          ++ showProtoId idt
+          ++ " to have the type "
+          ++ Show.shows rep ", instead it was set to "
+          ++ Show.shows val "."
+  {-# INLINE displayException #-}
+
+-- Proto Files -----------------------------------------------------------------
+
+-- | 'for'-style traversal over the 'ProtoServices' defined in a 'DotProto'.
+--
+-- @since 0.1.0.0
+servicesOf :: Applicative f => DotProto -> (ProtoService -> f a) -> f [a]
+servicesOf dotproto k = traverse k (getFileServices dotproto)
+
+-- | Accesses the list of services defined in the 'DotProto'.
+--
+-- @since 0.1.0.0
+getFileServices :: DotProto -> [ProtoService]
+getFileServices dotproto =
+  foldMap toService (DotProto.protoDefinitions dotproto)
+  where
+    toService :: DotProtoDefinition -> [ProtoService]
+    toService (DotProto.DotProtoService _ idt ps) = [ProtoService idt ps]
+    toService _ = []
+
+-- | Accesses the 'ProtoOptionSet' of options set in the 'DotProto'.
+--
+-- If the 'DotProto' sets the same file-level option more than once, a
+-- 'ProtoOptionAlreadySet' error is raised.
+--
+-- @since 0.1.0.0
+getFileOptions :: MonadError ProtoOptionError m => DotProto -> m ProtoOptionSet
+getFileOptions dotproto = foldrM insertOption mempty (DotProto.protoOptions dotproto)
+
+-- Proto Services --------------------------------------------------------------
+
+-- | 'ProtoService' is a record representing a protobuf service definition.
+--
+-- @since 0.1.0.0
+data ProtoService = ProtoService
+  { serviceName :: DotProtoIdentifier
+  , serviceDefs :: [DotProtoServicePart]
+  }
+  deriving (Eq, Show)
+
+-- | 'for'-style traversal over the 'RPCMethod' defined in a 'ProtoService'.
+--
+-- @since 0.1.0.0
+methodsOf :: Applicative f => ProtoService -> (RPCMethod -> f a) -> f [a]
+methodsOf dotproto k = traverse k (getServiceMethods dotproto)
```

Comment:
Just a style nit/suggestion, feel free to ignore because I think this is fine:

```suggestion
methodsOf dotproto = for (getServiceMethods dotproto)
```